src/mip/utils/timestamp.hpp

```diff
@@ -0,0 +1,222 @@
+#pragma once
+
+#include <assert.h>
+#include <chrono>
+
+using std::chrono::duration_cast;
+using Nanoseconds = std::chrono::nanoseconds;
+using Microseconds = std::chrono::microseconds;
+using Milliseconds = std::chrono::milliseconds;
+using Seconds = std::chrono::seconds;
+using Minutes = std::chrono::minutes;
+using Hours = std::chrono::hours;
+
+#if _HAS_CXX20
+    using Days = std::chrono::days;
+    using Weeks = std::chrono::weeks;
+    using Months = std::chrono::months;
+    using Years = std::chrono::years;
+#else
+    #include <ratio>
+    using Days = std::chrono::duration<int, std::ratio<86400>>;
+    using Weeks = std::chrono::duration<int, std::ratio<604800>>;
+    using Months = std::chrono::duration<int, std::ratio<2629746>>;
+    using Years = std::chrono::duration<int, std::ratio<31556952>>;
+#endif // _HAS_CXX20
+
+namespace mip
+{
+    /// -----------------------------------------------------
+    /// Legend:
+    /// -----------------------------------------------------
+    /// T --> Arithmetic type
+    /// D --> Duration (currently std::chrono::duration only)
+    /// -----------------------------------------------------
+    
+
+    /// Manages a timestamp that increments based on an underlying time system.
+    ///
+    /// Notes:
+    /// ---------------------------------------------------------------------------------
+    ///     * Currently supports GPS clock (only) for the underlying time system.
+    ///     * Might add support for different time systems in the future.
+    ///     * Currently supports std::chrono:duration (only) for timestamp inputs.
+    /// ---------------------------------------------------------------------------------
+    class TimestampState
+    {
+    public:
+        TimestampState() {}
+
+        /// Synchronizes timestamp to the current time in the underlying time system.
+        void synchronize();
```

Comment:
Not sold on `synchronize` for the name of this, just because there are so many potential time sources (other sensors, the pps, devices time, time from a GPS receiver, etc.).
Maybe `syncToSystemTime` or `updateToSystemTime`? The fact that it's GPS might also be good to specify here though too.

Reply:
My vision for this was to have the class handle different time sources. So you specify the time source when instantiating. Right now, it only has GPS time, but we could expand it if needed. So synchronize() would synchronize to whatever the underlying time source is set to. What do you think about that?

Could also name it something like syncToSource or syncToTimeSource?

Reply:
Changed the name to synchronizeGPS. Figured that made more sense than parameterizing a single function, because then that function has to parse the parameters and then call another private function anyway. Now I'm envisioning having a different synchronize function for each time system that can be called at any time. For example: synchronizeToGPS(), synchronizeToUTC(), etc. What do you think about this?

Reply:
Added a time standard parameter and set the default to GPS. How does that look?